[6.x] Core/Items: fix ItemUpgrade.

frozen4 opened the pull request:
**Changes proposed**:
- fix ItemUpgrade

**Target branch(es)**: 6x

**Issues addressed**: Fixes #

**Tests performed**:  tested in-game

**Known issues and TODO list**:
- [ ] After upgrade a item  you can't see item's stats change till you move the item to other slot
- [ ]  change the item's stats bonus

frozen4 committed:
Try to fix ItemUpgrade..

tkrokli commented:
PR Title suggestion: **[6.x] Core/DB/Hotfixes: implement ItemUpgrade**

frozen4 committed:
Add sql File

Signed-off-by: frozen4 <frozen4@163.com>

Lordron commented on src/server/game/DataStores/DB2Structure.h:
```diff
@@ -668,6 +668,24 @@ struct ItemBonusTreeNodeEntry
+struct RulesetItemUpgradeEntry
+{
+    uint32 ID;
+    uint32 unk;
```
index

DDuarte commented on src/server/game/Server/Packets/ItemPackets.h:
```diff
@@ -502,6 +502,20 @@ namespace WorldPackets
+            ObjectGuid ItemGUID;
+            uint32 UpgradeID;
+            uint32 bag;
+            uint32 slot;
```
Replace all fields by

```c++
        ObjectGuid ItemMaster;
        ObjectGuid ItemGUID;
        int32 ContainerSlot = 0;
        int32 UpgradeID = 0;
        int32 Slot = 0;
```

frozen4 replied:
ObjectGuid ItemMaster;
why  ItemMaster? it is Npc's GUID

DDuarte replied:
Official names.

DDuarte commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -11038,6 +11038,9 @@ Item* Player::StoreNewItem(ItemPosCountVec const& pos, uint32 itemId, bool updat
         if (randomPropertyId)
             item->SetItemRandomProperties(randomPropertyId);
 
+        if (uint32 UpgradeID = sDB2Manager.GetRulesetItemUpgrade(itemId))
```
Local variables should be lower camel case -> `upgradeId`

DDuarte commented on src/server/game/DataStores/DB2fmt.h:
```diff
@@ -62,6 +62,8 @@ char const ImportPriceWeaponFormat[] = "nf";
 char const ItemBonusFormat[] = "niiiii";
 char const ItemBonusTreeNodeFormat[] = "niiii";
+const char ItemUpgradeEntryfmt[] = "niiiii";
+const char RulesetItemUpgradeEntryfmt[] = "niii";
```
`char const`, like all the other ones

Shauren replied:
And the suffix must be "Format", not "fmt"

frozen4 committed:
index

Signed-off-by: frozen4 <frozen4@163.com>

Shauren commented on src/server/database/Database/Implementation/HotfixDatabase.h:
```diff
@@ -142,6 +142,9 @@ enum HotfixDatabaseStatements
     HOTFIX_SEL_ITEM_CLASS_LOCALE,
 
+    HOTFIX_SEL_ITEM_UPGRADE,
+    HOTFIX_SEL_RULESET_ITEM_UPGRADE,
```
Formatting: sort by db2 name - RULESET_* doesn't belong here

Shauren commented on src/server/game/DataStores/DB2Stores.cpp:
```diff
@@ -250,6 +252,8 @@ void DB2Manager::LoadStores(std::string const& dataPath, uint32 defaultLocale)
     LOAD_DB2(sItemBonusStore);
     LOAD_DB2(sItemBonusTreeNodeStore);
+    LOAD_DB2(sItemUpgradeStore);
+    LOAD_DB2(sRulesetItemUpgradeStore);
```
Formatting: sort by db2 name

Shauren commented on src/server/game/DataStores/DB2Stores.h:
```diff
@@ -150,6 +151,7 @@ class TC_GAME_API DB2Manager
     typedef std::unordered_multimap<uint32 /*itemId*/, uint32 /*bonusTreeId*/> ItemToBonusTreeContainer;
     typedef std::unordered_map<uint32 /*itemId | appearanceMod << 24*/, uint32> ItemDisplayIdContainer;
     typedef std::unordered_map<uint32, std::set<ItemBonusTreeNodeEntry const*>> ItemBonusTreeContainer;
+    typedef std::map<uint32, uint32> RulesetItemUpgradeContainer;
```
Formatting: sort by db2 name

Shauren commented on src/server/game/DataStores/DB2Stores.h:
```diff
@@ -181,6 +183,7 @@ class TC_GAME_API DB2Manager
     uint32 GetHeirloomItemLevel(uint32 curveId, uint32 level) const;
     HeirloomEntry const* GetHeirloomByItemId(uint32 itemId) const;
     ItemBonusList const* GetItemBonusList(uint32 bonusListId) const;
+    uint32 GetRulesetItemUpgrade(uint32 itemId);
```
Formatting: sort by db2 name

Shauren commented on src/server/game/DataStores/DB2Structure.h:
```diff
@@ -668,6 +668,24 @@ struct ItemBonusTreeNodeEntry
+    uint32 ID;                                                      // 0
+    uint32 ItemUpgradePath;                                     // 1
+    uint32 LevelUpgrade;                                          // 2
+    uint32 prevItemUpgradeId;                                 // 3
```
All fields must be UpperCamelCase

Shauren commented on src/server/game/DataStores/DB2Structure.h:
```diff
@@ -668,6 +668,24 @@ struct ItemBonusTreeNodeEntry
+    uint32 ItemUpgradeID;                                    // 2
+    uint32 ItemID;                                               // 3
+};
+
```
Formatting: comment alignment

Shauren commented on src/server/game/DataStores/DB2Structure.h:
```diff
@@ -668,6 +668,24 @@ struct ItemBonusTreeNodeEntry
+struct RulesetItemUpgradeEntry
+{
+    uint32 ID;                                                      // 0
+    uint32 unk;                                                    // 1
```
All fields must be UpperCamelCase

frozen4 committed:
DB/Misc: Change file name

DDuarte commented:
>  After upgrade a item you can't see item's stats change till you move the item to other slot

Have you tried sending `ClientItemUpgradeResult`?

frozen4 committed:
weird rule

frozen4 committed:
ok,  send ItemUpgradeResult
but item's stats still not update

Lordron commented on src/server/game/DataStores/DB2Structure.h:
```diff
@@ -673,15 +673,15 @@ struct ItemUpgradeEntry
 {
     uint32 ID;                                                      // 0
-    uint32 unk;                                                    // 1
+    uint32 Unk;                                                    // 1
```
Unk -> Index

frozen4 committed:
All fields must be UpperCamelCase

frozen4 commented:
it is so tired to follow the code style.
I also fix the Garrison's mission.
but I think if me commit it 
it will be many break rules code

sorry my bad English....

Ladel commented on src/server/game/Server/Packets/ItemPackets.cpp:
```diff
@@ -529,6 +529,15 @@ void WorldPackets::Item::UseCritterItem::Read()
+void WorldPackets::Item::UpgradeItem::Read()
+{
+    _worldPacket >> Npc;
+    _worldPacket >> ItemGuid;
```
ItemGUID

Shauren commented on src/server/game/Server/Packets/ItemPackets.cpp:
```diff
@@ -544,3 +553,11 @@ WorldPacket const* WorldPackets::Item::SocketGemsResult::Write()
+WorldPacket const* WorldPackets::Item::ItemUpgradeResult::Write()
+{
+    _worldPacket.WriteBit(Success);
+    _worldPacket.FlushBits();//dose it needed?
```
to answer this question - yes, you must call FlushBits if writing a bit is the last thing done in a packet (and remove the comment since I just answered it)

frozen4 committed:
mistake

Shauren commented on src/server/game/Server/Protocol/Opcodes.h:
```diff
@@ -1151,6 +1151,7 @@ enum OpcodeServer : uint32
     SMSG_ITEM_PURCHASE_REFUND_RESULT                  = 0x2535,
     SMSG_ITEM_PUSH_RESULT                             = 0x25BA,
     SMSG_ITEM_TIME_UPDATE                             = 0x2714,
+    SMSG_ITEM_UPGRADE_RESULT                     = 0x25DB,
```
0x25DB does not have any handler in 6.2.4 client - sending it will not affect anything
Since it does not have a handler, please place this line at the bottom of the enum, just after SMSG_ACCOUNT_HEIRLOOM_UPDATE. Reason for this is that the entire enum of opcodes that have client handlers is generated automatically on every patch so anything unhandled there will simply get lost.

frozen4 committed:
Update ItemPackets.cpp

frozen4 committed:
Update Opcodes.h


ClientToServer: CMSG_UPGRADE_ITEM (0x3187) Length: 34 ConnIdx: 1 Time: 06/21/2016 02:06:24.877 Number: 128
|-------------------------------------------------|---------------------------------|
| 00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F | 0 1 2 3 4 5 6 7 8 9 A B C D E F |
|-------------------------------------------------|---------------------------------|
| 27 E6 CE E5 61 29 5D 43 18 21 20 9F E0 65 77 E9 | ' . . . a ) ] C . !   . . e w . |
| 09 03 40 18 21 0C FB 01 00 00 FF 00 00 00 04 00 | . . @ . ! . . . . . . . . . . . |
| 00 00                                           | . .                             |
|-------------------------------------------------|---------------------------------|

ServerToClient: SMSG_UPDATE_OBJECT (0x277D) Length: 483 ConnIdx: 1 Time: 06/21/2016 02:06:24.877 Number: 130
NumObjUpdates: 1
MapID: 0 (0)
HasDestroyObjects: False
Data size: 472
[0] UpdateType: Values
[0] Object Guid: Full: 0x08211800000000000000000003F3A71C Player/0 R2118/S0 Map: 0 Low: 66299676
[0] UNIT_FIELD_HEALTH: 232200/3.253815E-40

ServerToClient: SMSG_SET_CURRENCY (0x24F1) Length: 13 ConnIdx: 1 Time: 06/21/2016 02:06:25.033 Number: 131
Type: 738
Quantity: 475
Flags: 0
HasTrackedQuantity: False
HasWeeklyQuantity: False
SuppressChatLog: False

ServerToClient: SMSG_ITEM_EXPIRE_PURCHASE_REFUND (0x2537) Length: 11 ConnIdx: 1 Time: 06/21/2016 02:06:25.033 Number: 132
ItemGUID: Full: 0x0C211800000000004000000309E97765 Item/0 R2118/S4194304 Map: 0 Low: 13051197285

ServerToClient: 9691 (0x25DB) Length: 1 ConnIdx: 1 Time: 06/21/2016 02:06:25.033 Number: 267
|-------------------------------------------------|---------------------------------|
| 00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F | 0 1 2 3 4 5 6 7 8 9 A B C D E F |
|-------------------------------------------------|---------------------------------|
| 80                                              | .                               |
|-------------------------------------------------|---------------------------------|

tkrokli commented:
It is good that you contribute. Don't feel bad about your English skill, we can help you with that by posting suggestions for text improvements, if and when it is needed.

DDuarte commented on src/server/game/Server/Packets/ItemPackets.h:
```diff
@@ -502,6 +502,30 @@ namespace WorldPackets
+
+            void Read() override;
+
+            ObjectGuid Npc;
```
`ItemMaster` instead of `Npc`

Shauren commented:
Merged, thanks

P-Kito commented:
Thanks @frozen4

tkrokli commented:
Thank you for this PR contribution, @frozen4 . :+1: As you probably will notice in the finished commit, you will see that Shauren has fixed the code spacing and the db2 sorting he commented on in your PR. :-)
